telemetry/run.py

```diff
@@ -0,0 +1,99 @@
+import subprocess
+import threading
+import argparse
+import datetime
+import sys
+import os
+
+
+def parse_cpu_stat(info:list) -> dict:
+    '''
+    Takes data returned by sensors, which contains temperature / voltage data.
+    Returns a dict, whose keys are (group) name of the sensors and values are pairs of (label x data).
+    '''
+    current_group = 'ungrouped'
+    result = dict({ current_group : [] })
+    for line in info:
+        # omit empty lines and `N/A` data
+        if line and not 'N/A' in line:
+            # Adapter names does not contain spaces
+            if ' ' not in line:
+                current_group = line
+                result.update({ line : [] })
+            else:
+                label, data = ' '.join(line.split()).split()[:2]
+                result[current_group].append((label.replace(':', ''), ''.join(data)))
+    # return with empty entries removed
+    return dict(filter(lambda x: len(x[1]), result.items()))
+
+def parse_gpu_stat(info:list) -> list:
+    '''
+    Takes data returned by `nvidia-smi`, which contains a string with two lines, returns
+    the last line, which is the data we need to monitor during an experiment.
+    '''
+    try:
+        # ensure the list is not empty
+        assert info
+    except:
+        return []
+    return info[-1].decode().split(', ')
+
+def start_job(fp_dir, nvidia_fields, time_span, time_run) -> None:
+    '''
+    A chornological job, runs every `time_span` seconds.
+    Fetches data from `nvidia-smi` and `sensors`, then parse it
+    and write to corresponding files. Data will not be processed
+    during monitoring in order to minimize the influence to the running experiments.
+
+    Note: The process will be halted by `dashboard.py` when an experiment ends. 
+    '''
+    threading.Timer(time_span, start_job, args=[fp_dir, nvidia_fields, time_span, time_run + 1]).start()
+    nvidia_smi = subprocess.Popen(['nvidia-smi', '--format=csv', '--query-gpu={}'.format(','.join(nvidia_fields))], stdout=subprocess.PIPE)
```

Comment:
Can you not use `check_output` in `subprocess` instead? Apparently `Popen` requires explicitly waiting for and cleaning up the child processes, which is why this was leaving zombie processes

Reply:
Changed to `subprocess.run` to deal with timeout. This may avoid creating zombie processes.